core/src/main/java/org/opensearch/sql/executor/pagination/PaginatedPlanCache.java

```diff
@@ -30,132 +30,99 @@
 @RequiredArgsConstructor
 public class PaginatedPlanCache {
   public static final String CURSOR_PREFIX = "n:";
-  private final StorageEngine storageEngine;
+
+  private final StorageEngine engine;
 
   public boolean canConvertToCursor(UnresolvedPlan plan) {
     return plan.accept(new CanPaginateVisitor(), null);
   }
 
   /**
-   * Converts a physical plan tree to a cursor. May cache plan related data somewhere.
+   * Converts a physical plan tree to a cursor.
    */
-  public Cursor convertToCursor(PhysicalPlan plan) throws IOException {
-    if (plan instanceof PaginateOperator) {
-      var cursor = plan.toCursor();
-      if (cursor == null) {
-        return Cursor.None;
-      }
-      var raw = CURSOR_PREFIX + compress(cursor);
-      return new Cursor(raw.getBytes());
+  public Cursor convertToCursor(PhysicalPlan plan) {
+    try {
+      return new Cursor(CURSOR_PREFIX
+          + serialize(((SerializablePlan) plan).getPlanForSerialization()));
+      // ClassCastException thrown when a plan in the tree doesn't implement SerializablePlan
+    } catch (ClassCastException | NoCursorException e) {
```

Comment:
When can `ClassCastException` be thrown?

Reply:
We are always trying to convert a plan tree to a cursor in https://github.com/Bit-Quill/opensearch-project-sql/blob/10ba1d66d663113a70ed227762732944c178f491/opensearch/src/main/java/org/opensearch/sql/opensearch/executor/OpenSearchExecutionEngine.java#L55
If tree contains a plan which doesn't implement `SerializablePlan`, there would be a `ClassCastException` somewhere, e.g. in https://github.com/Bit-Quill/opensearch-project-sql/blob/10ba1d66d663113a70ed227762732944c178f491/opensearch/src/main/java/org/opensearch/sql/opensearch/executor/protector/ResourceMonitorPlan.java#L93-L96

Reply:
I don't serialize `PaginateOperator` (to reduce cursor size), so I can't check `plan instanceof PaginateOperator` here as it was before.
https://github.com/Bit-Quill/opensearch-project-sql/blob/0ff3390447831bc49c0c625a9097d85a364a9599/core/src/main/java/org/opensearch/sql/planner/physical/PaginateOperator.java#L69-L73
That means I have to do another protection here. The incoming tree could be `ProjectOperator -> ResourceMonitorPlan -> PrometheusIndexScan` or `ProjectOperator -> ValuesOperator`.